### ISSUE TITLE
aws_eks_cluster and create_before_destroy

### ISSUE THREAD

**Issue by tristanpemble** [hashicorp/terraform-provider-aws#4771]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.aws v1.22.0
```

### Affected Resource(s)

* aws_eks_cluster

### Description

Amazon requires EKS clusters to have a unique name. This means `create_before_destroy` cannot be used on an EKS cluster, because you will get a `ResourceInUseException: Cluster already exists with name` whenever there are changes to the cluster.

This is unfortunate, since in order to configure an ASG for worker nodes, you need to reference `aws_eks_cluster.endpoint` in the user data of the launch configuration. In order to use launch configurations effectively, you have to use `create_before_destroy = true` on them.

And since the launch configuration depends on the EKS cluster, that forces the EKS cluster to also `create_before_destroy`. If you have changes to both launch configuration and cluster, then `terraform apply` will fail without a clear fix.

I'm not sure what solution there is or if there is any, but I wanted to open this to have a discussion about the issue.

**Comment by SpamapS** [comment]
Yep, I've run into this too. The only way I can find to change an EKS that uses ASG's with terraform is to rip the whole thing down with destroy.

**Comment by nhuray** [comment]
I've run into this too...I finally remove manually the cluster then refresh and apply.

**Comment by jamesdobson** [comment]
Now that the userdata no longer needs a reference to the cluster endpoint URL, do you still get this problem?

See the launch configuration in the new CloudFormation template: https://amazon-eks.s3-us-west-2.amazonaws.com/cloudformation/2018-08-30/amazon-eks-nodegroup.yaml

For reference, the bootstrap script that is now baked in to the AMIs is here: https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh

**Comment by aaron-trout** [comment]
Oh man, thanks so much for this! I never would have figured it out! 😆 

Yeah, this is pretty annoying! What is more annoying though is the massive inconsistency in AWS where some resources have name as a first class citizen and some do not, and for some it must be unique... and some not. WTF Amazon!?

**Comment by tdmalone** [comment]
Thinking out loud here... would a possible workaround be interpolating a random UUID in to the cluster name? See https://www.terraform.io/docs/providers/random/r/uuid.html

**Comment by th31nitiate** [comment]
> Oh man, thanks so much for this! I never would have figured it out! 😆
> 
> Yeah, this is pretty annoying! What is more annoying though is the massive inconsistency in AWS where some resources have named as a first-class citizen and some do not, and for some, it must be unique... and some not. WTF Amazon!?

Those operations and developer silos are the main cause of this. This could also be an onboarding problem but I was going to blame terraform but turns out is an Amazon thing. I was going to suggest to just use cloud formation but that is not an option in this scenario. I can see where data issues start to happen, how. 

> Thinking out loud here... would a possible workaround be interpolating a random UUID into the cluster name? See https://www.terraform.io/docs/providers/random/r/uuid.html

You can do, add a prefix to the cluster name. The only problem is that may create a new cluster for each randomly generated URL.

**Comment by jamesongithub** [comment]
Seems like `aws_eks_cluster` and `aws_eks_node_group` should add support for `name_prefix`.

**Comment by bflad** [comment]
There are now instructions about how to [add naming generation to resources in the Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#adding-resource-name-generation-support).

**Comment by schollii** [comment]
> There are now instructions about how to [add naming generation to resources in the Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#adding-resource-name-generation-support).

@bflad The link does not work, could you update it please?

**Comment by mattburgess** [comment]
As per the [docs](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html) - versions of the AWS Load Balancer Controller after 2.1.1 (latest is 2.4.5) no longer require subnets to be tagged so my [comment](https://github.com/hashicorp/terraform-provider-aws/issues/4771#issuecomment-940384970) is no longer accurate. I reckon it may be possible for `name_prefix` support to be implemented now as per the [guide](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/resource-name-generation.md).

**Comment by bryantbiggs** [comment]
@tristanpemble is this issue still relevant or can we close it out?

**Comment by tristanpemble** [comment]
@bryantbiggs I haven't touched Terraform + Kubernetes in many years, so I wouldn't know. you'd have to ask the crowd to be sure.
